### ISSUE TITLE
Pkg.test: ERROR: Package ClimaCore errored during testing (received signal: KILL)

### ISSUE THREAD

**Issue by simonbyrne** [JuliaLang/julia#47890]
When running tests with coverage (`Pkg.test(coverage=true)`) we see the following error:
https://github.com/CliMA/ClimaCore.jl/actions/runs/3687762429/jobs/6241745116#step:7:480

```
ERROR: Package ClimaCore errored during testing (received signal: KILL)
Stacktrace:
 [1] pkgerror(msg::String)
   @ Pkg.Types /opt/hostedtoolcache/julia/1.8.3/x64/share/julia/stdlib/v1.8/Pkg/src/Types.jl:67
 [2] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, julia_args::Cmd, test_args::Cmd, test_fn::Nothing, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
   @ Pkg.Operations /opt/hostedtoolcache/julia/1.8.3/x64/share/julia/stdlib/v1.8/Pkg/src/Operations.jl:1813
 [3] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, test_fn::Nothing, julia_args::Cmd, test_args::Cmd, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool, kwargs::Base.Pairs{Symbol, Base.PipeEndpoint, Tuple{Symbol}, NamedTuple{(:io,), Tuple{Base.PipeEndpoint}}})
   @ Pkg.API /opt/hostedtoolcache/julia/1.8.3/x64/share/julia/stdlib/v1.8/Pkg/src/API.jl:434
 [4] test(pkgs::Vector{Pkg.Types.PackageSpec}; io::Base.PipeEndpoint, kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:coverage,), Tuple{Bool}}})
   @ Pkg.API /opt/hostedtoolcache/julia/1.8.3/x64/share/julia/stdlib/v1.8/Pkg/src/API.jl:156
 [5] test(; name::Nothing, uuid::Nothing, version::Nothing, url::Nothing, rev::Nothing, path::Nothing, mode::Pkg.Types.PackageMode, subdir::Nothing, kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:coverage,), Tuple{Bool}}})
   @ Pkg.API /opt/hostedtoolcache/julia/1.8.3/x64/share/julia/stdlib/v1.8/Pkg/src/API.jl:171
 [6] top-level scope
   @ none:1
Error: Process completed with exit code 1.
```
This was with julia version 1.8.3, running on GitHub Actions using Ubuntu 20.04.

It's not clear what is causing this, or when in the test it occurs (the stacktrace doesn't give any useful information), but it is seen elsewhere (not all of these appear to involve coverage), but they do appear to be in Julia 1.8:
- https://github.com/julia-actions/julia-runtest/issues/59
- https://github.com/Circuitscape/Circuitscape.jl/issues/364
- https://github.com/SciML/DiffEqBase.jl/pull/803#issuecomment-1219359987
- https://github.com/SciML/SciMLSensitivity.jl/pull/570
- https://github.com/Nemocas/AbstractAlgebra.jl/pull/1145#issuecomment-1117311492

cc: @DilumAluthge @ranjanan @ChrisRackauckas

**Comment by DilumAluthge** [comment]
@simonbyrne Just to clarify, when you do `Pkg.test(coverage=false)`, you do not see this error, right?

**Comment by simonbyrne** [comment]
I hadn't seen it before, but it's hard to say,.

**Comment by simonbyrne** [comment]
> @simonbyrne Just to clarify, when you do `Pkg.test(coverage=false)`, you do not see this error, right?

We do not, but https://github.com/Circuitscape/Circuitscape.jl/issues/364 appears to be run without coverage

**Comment by DilumAluthge** [comment]
Are you able to reproduce this locally? If so, could you get an rr trace?

**Comment by DilumAluthge** [comment]
Also, if you are able to reproduce this locally, do any of the system logs give any insight as to why the process was killed?

**Comment by simonbyrne** [comment]
No, can only repro it so far on GitHub Actions. And I don't have access to a box that can run rr

**Comment by jpsamaroo** [comment]
`SIGKILL` is usually OOM, have you checked how much memory the process uses before it's killed? An OOM will probably also be reported in `dmesg`, IIRC.

**Comment by simonbyrne** [comment]
Ah, ok: so what I think is happening is that the new process Pkg.test launches is being OOM-killed (and doesn't get a stacktrace because SIGKILL won't let it exit gracefully): this process presumably exits with 137 (128 + 9). The outer process just sees this as a test failure, and so exits with error code 1.

**Comment by IanButterworth** [comment]
I had a similar situation with PackageCompiler where the child process was OOM killed, and the error served wasn't helpful.

As a UX patch for PackageCompiler, I added [a memory monitor ](https://github.com/JuliaLang/PackageCompiler.jl/pull/740) that adds a hint if the child has crashed and memory load was very high during its existence. Ideally there would be some OOM-kill indication from the OS, but that doesn't seem to be the case, and searching through system logs seems too brittle/difficult cross-platform.

I guess a similar monitor could be added to `Pkg.test` but I don't know whether running more things async alongside tests is a good idea.
